### PR TITLE
Add support for decimal values to Variant

### PR DESCRIPTION
**Pull request by majetideepak**

### CODE REVIEW

File: velox/type/tests/VariantTest.cpp [hunk 1]
```diff
@@ -89,3 +89,24 @@ TEST(Variant, opaque) {
...
   }
 }
+
+TEST(Variant, decimal) {
+  {
+    auto shortDecimalType = DECIMAL(10, 4);
+    variant v = variant::shortDecimal(100, shortDecimalType);
+    EXPECT_TRUE(v.isSet());
+    EXPECT_EQ(TypeKind::SHORT_DECIMAL, v.kind());
+    EXPECT_EQ(100, v.value<TypeKind::SHORT_DECIMAL>().value.unscaledValue());
```
Comment (mbasmanova) [inline velox/type/tests/VariantTest.cpp:99]: Would you also check precision and scale?

File: velox/type/tests/VariantTest.cpp [hunk 2]
```diff
@@ -89,3 +89,24 @@ TEST(Variant, opaque) {
...
+    EXPECT_EQ(*v.inferType(), *shortDecimalType);
+    EXPECT_EQ(v.toJson(), "0.0100");
```
Comment (mbasmanova) [inline velox/type/tests/VariantTest.cpp:101]: Would you also add tests for comparing decimals and making null decimals?

File: velox/type/Variant.h [hunk 3]
```diff
@@ -108,6 +115,44 @@ struct VariantTypeTraits<TypeKind::ARRAY> {
   using stored_type = std::vector<variant>;
 };
 
+template <typename T>
+struct DecimalCapsule {
```
Comment (mbasmanova) [inline velox/type/Variant.h:119]: Would you add some comments to explain the purpose of this struct?

**Review by mbasmanova — COMMENTED** [review]
@majetideepak Look great % a couple of minor comments.

File: velox/type/Variant.h [hunk 4]
```diff
@@ -301,6 +347,40 @@ class variant {
...
+    auto decimalType = type->asShortDecimal();
+    if (input.has_value()) {
+      return {
+          TypeKind::SHORT_DECIMAL,
+          new detail::ShortDecimalCapsule{
+              ShortDecimal(input.value()),
+              decimalType.precision(),
+              decimalType.scale()}};
+    } else {
+      return {TypeKind::SHORT_DECIMAL, nullptr};
```
Comment (mbasmanova) [inline velox/type/Variant.h:363]: Do we need to make some changes to block usage of `variant::null(TypeKind::SHORT_DECIMAL)`?

```
  static variant null(TypeKind kind) {
    return variant{kind};
  }
```

File: velox/type/tests/VariantTest.cpp [hunk 5]
```diff
@@ -89,3 +89,53 @@ TEST(Variant, opaque) {
...
+  EXPECT_EQ(TypeKind::SHORT_DECIMAL, v.kind());
+  EXPECT_EQ(1234, v.value<TypeKind::SHORT_DECIMAL>().value.unscaledValue());
+  EXPECT_EQ(10, v.value<TypeKind::SHORT_DECIMAL>().precision);
+  EXPECT_EQ(3, v.value<TypeKind::SHORT_DECIMAL>().scale);
+  EXPECT_EQ(*v.inferType(), *shortDecimalType);
+  EXPECT_EQ(v.toJson(), "1.234");
+  // 1.2345
+  variant u1 = variant::shortDecimal(12345, DECIMAL(10, 4));
+  // 1.2345 > 1.234
+  EXPECT_FALSE(
```
Comment (mbasmanova) [inline velox/type/tests/VariantTest.cpp:106]: EXPECT_LT ?

File: velox/type/Variant.h (hunk 4 above)
Comment (majetideepak) [reply velox/type/Variant.h:363]: Good catch! I missed this.

File: velox/type/Variant.h [hunk 6]
```diff
@@ -108,6 +115,50 @@ struct VariantTypeTraits<TypeKind::ARRAY> {
...
 
+/// Variants contain a TypeKind and a value.
+/// DECIMAL type variants use TypeKind to store the kind, and this struct as the
+/// value to store the unscaled value, precision, scale, and null information.
+template <typename T>
+struct DecimalCapsule {
+  T value;
+  int precision;
+  int scale;
+  bool isNull = false;
```
Comment (majetideepak) [inline velox/type/Variant.h:126]: @mbasmanova I added another field here to capture null information.

Comment (mbasmanova) [inline velox/type/Variant.h:126]: Any particular reason not to combine value and isNull into an std::optional<int>?

File: velox/type/Variant.h [hunk 7]
```diff
@@ -469,6 +545,15 @@ class variant {
     return value<CppToType<T>::typeKind>();
   }
 
+  bool isNull() const {
```
Comment (mbasmanova) [inline velox/type/Variant.h:548]: Any particular reason for moving this method? Looks like there are other methods that rely on ptr_ null-ness: isSet, checkNotNull, etc. Do these need updating as well?

File: velox/type/Variant.h (hunk 6 above)
Comment (majetideepak) [reply velox/type/Variant.h:126]: We can serialize a variant via variant::serialize(), (not implemented in this PR). It seemed to be that `std::optional` is not supported.

File: velox/type/Variant.h (hunk 7 above)
Comment (majetideepak) [reply velox/type/Variant.h:548]: They do indeed. Will fix.

Comment (majetideepak) [reply velox/type/Variant.h:548]: I had to move this below the `value()` implementation as it uses the `value()` method.

File: velox/type/Variant.h (hunk 6 above)
Comment (mbasmanova) [reply velox/type/Variant.h:126]: This is custom logic, no? What would prevent us from supporting an optional there?

File: velox/type/Variant.h (hunk 7 above)
Comment (majetideepak) [reply velox/type/Variant.h:544]: I removed `isSet` since there is a `hasValue()` with the same implementation.

File: velox/type/Variant.h (hunk 6 above)
Comment (majetideepak) [reply velox/type/Variant.h:126]: You are right. I will fix this.

File: velox/type/tests/VariantTest.cpp [hunk 8]
```diff
@@ -89,3 +89,57 @@ TEST(Variant, opaque) {
...
+  variant u1 = variant::shortDecimal(12345, DECIMAL(10, 4));
+  // 1.2345 > 1.234
+  EXPECT_LT(
+      v.value<TypeKind::SHORT_DECIMAL>(), u1.value<TypeKind::SHORT_DECIMAL>());
+  // 0.1234
+  variant u2 = variant::shortDecimal(1234, DECIMAL(10, 4));
+  // 0.1234 < 1.234
+  EXPECT_LT(
+      u2.value<TypeKind::SHORT_DECIMAL>(), v.value<TypeKind::SHORT_DECIMAL>());
+  variant null = variant::shortDecimal(std::nullopt, DECIMAL(10, 5));
```
Comment (mbasmanova) [inline velox/type/tests/VariantTest.cpp:113]: nit: perhaps, extract this logic into a separate test or separate it with a comment for readability

File: velox/type/Variant.h [hunk 9]
```diff
@@ -108,6 +115,58 @@ struct VariantTypeTraits<TypeKind::ARRAY> {
...
+struct DecimalCapsule {
+  std::optional<T> unscaledValue;
+  int precision;
+  int scale;
+
+  T value() const {
+    return unscaledValue.value();
+  }
+
+  bool has_value() const {
```
Comment (mbasmanova) [inline velox/type/Variant.h:131]: naming: hasValue

**Review by mbasmanova — COMMENTED** [review]
@majetideepak 

Looks like there might be a few more places that need updating.

```
  variant() : kind_{TypeKind::INVALID}, ptr_{nullptr} {}

  variant(TypeKind kind) : kind_{kind}, ptr_{nullptr} {}
```

File: velox/type/Variant.h (hunk 9 above)
Comment (majetideepak) [reply velox/type/Variant.h:131]: good catch!

File: velox/type/tests/VariantTest.cpp (hunk 8 above)
Comment (majetideepak) [reply velox/type/tests/VariantTest.cpp:113]: moved the logic as separate tests

**Comment by majetideepak** [comment]
@mbasmanova the previous commit seems to cause some failures in the internal ci.  Let me know if I can help in any way. Thanks.

**Comment by mbasmanova** [comment]
> @mbasmanova the previous commit seems to cause some failures in the internal ci. Let me know if I can help in any way. Thanks.

I'm working though these failures. Some of these are due to removed isSet method. Would you add it back?

```
  bool isSet() const {
    return !isNull();
  }
```

**Comment by majetideepak** [comment]
> I'm working though these failures. Some of these are due to removed isSet method. Would you add it back?

done.

File: velox/type/Variant.h [hunk 10]
```diff
@@ -398,9 +398,17 @@ class variant {
             std::move(inputs)}};
   }
 
-  variant() : kind_{TypeKind::INVALID}, ptr_{nullptr} {}
+  variant() : kind_{TypeKind::INVALID}, ptr_{nullptr} {
+    VELOX_CHECK(
```
[the remark this answers is not in the document]
Comment (majetideepak) [reply velox/type/Variant.h]: Yes, indeed!

**Review by mbasmanova — APPROVED** [review]

File: velox/type/Variant.h [hunk 11]
```diff
@@ -108,6 +115,58 @@ struct VariantTypeTraits<TypeKind::ARRAY> {
...
+        (value().unscaledValue() / DecimalUtil::kPowersOfTen[scale]);
+    auto rhsIntegral =
+        (other.value().unscaledValue() /
+         DecimalUtil::kPowersOfTen[other.scale]);
+    if (lhsIntegral == rhsIntegral) {
+      return (value().unscaledValue() % DecimalUtil::kPowersOfTen[scale]) <
+          (other.value().unscaledValue() %
+           DecimalUtil::kPowersOfTen[other.scale]);
+    }
+    return lhsIntegral < rhsIntegral;
```
Comment (karteekmurthys) [inline velox/type/Variant.h:153]: Can't we leave it VELOX_UNSUPPORTED?

**Pull request closed without merging**